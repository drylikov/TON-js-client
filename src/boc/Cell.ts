import { Maybe } from '../types';
import { BitString } from './BitString';
import { deserializeBoc, hashCell, serializeToBoc } from './boc';
import inspectSymbol from 'symbol.inspect';
import { Slice } from '..';

export class Cell {

    static fromBoc(src: Buffer | string): Cell[] {
        return deserializeBoc(typeof src === 'string' ? Buffer.from(src, 'hex') : src);
    }

    readonly bits: BitString;
    readonly refs: Cell[] = [];
    readonly isExotic: boolean;

    constructor(isExotic: boolean = false, bits = BitString.alloc(1023)) {
        this.isExotic = isExotic;
        this.bits = bits;
    }

    beginParse() {
        return Slice.fromCell(this);
    }

    writeCell(anotherCell: Cell) {
        this.bits.writeBitString(anotherCell.bits);
        for (let r of anotherCell.refs) {
            this.refs.push(r);
        }
    }

    hash() {
        return hashCell(this);
    }

    toBoc(opts?: { idx?: Maybe<boolean>, crc32?: Maybe<boolean>, cacheBits?: Maybe<boolean>, flags?: Maybe<number> }) {
        let idx = (opts && opts.idx !== null && opts.idx !== undefined) ? opts.idx : true;
        let crc32 = (opts && opts.crc32 !== null && opts.crc32 !== undefined) ? opts.crc32 : true;
        let cacheBits = (opts && opts.cacheBits !== null && opts.cacheBits !== undefined) ? opts.cacheBits : false;
        let flags = (opts && opts.flags !== null && opts.flags !== undefined) ? opts.flags : 0;
        return serializeToBoc(this, idx, crc32, cacheBits, flags);
    }

    [inspectSymbol] = () => this.toString()

    toString(indent?: string): string {
        let id = indent || '';
        let s = id + 'x{' + this.bits.toFiftHex() + '}\n';
        for (let k in this.refs) {
            const i = this.refs[k];
            s += i.toString(id + ' ');
        }
        return s;
    }

    withReference(cell: Cell) {
        this.refs.push(cell);
        return this;
    }

    withData(src: string) {
        for (let s of src) {
            if (s === '0') {
                this.bits.writeBit(0);
            } else {
                this.bits.writeBit(1);
            }
        }
        return this;
    }

    equals(src: Cell) {
        if (src.refs.length !== this.refs.length) {
            return false;
        }
        for (let i = 0; i < src.refs.length; i++) {
            if (!src.refs[i].equals(this.refs[i])) {
                return false;
            }
        }
        return this.bits.equals(src.bits);
    }
}