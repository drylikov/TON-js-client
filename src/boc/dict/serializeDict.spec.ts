import { Cell } from "../Cell";
import { buildTree, serializeDict } from "./serializeDict";

describe('serializeDict', () => {
    it('should build prefix tree', () => {

        // From docs
        const map = new Map<string, number>();
        map.set('13', 169);
        map.set('17', 289);
        map.set('239', 57121);

        // Test built tree
        let tree = buildTree(map, 16);
        expect(tree).toMatchSnapshot();

        // Test serialization
        const res = serializeDict(map, 16, (src, cell) => cell.bits.writeUint(src, 16));
        let root = new Cell()
            .withData('11001000')
            .withReference(new Cell()
                .withData('011000')
                .withReference(new Cell()
                    .withData('1010011010000000010101001'))
                .withReference(new Cell()
                    .withData('1010000010000000100100001')))
            .withReference(new Cell()
                .withData('1011111011111101111100100001'));
        expect(res.equals(root)).toBe(true);
    });
});