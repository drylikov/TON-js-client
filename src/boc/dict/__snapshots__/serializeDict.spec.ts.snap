// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serializeDict should build prefix tree 1`] = `
Object {
  "label": "00000000",
  "node": Object {
    "left": Object {
      "label": "00",
      "node": Object {
        "left": Object {
          "label": "1101",
          "node": Object {
            "type": "leaf",
            "value": 169,
          },
        },
        "right": Object {
          "label": "0001",
          "node": Object {
            "type": "leaf",
            "value": 289,
          },
        },
        "type": "fork",
      },
    },
    "right": Object {
      "label": "1101111",
      "node": Object {
        "type": "leaf",
        "value": 57121,
      },
    },
    "type": "fork",
  },
}
`;
